File: web/src/main/java/by/sorface/sso/web/config/security/jwt/IdTokenCustomizerConfig.java
package by.sorface.sso.web.config.security.jwt;

import by.sorface.sso.web.records.principals.DefaultPrincipal;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.oauth2.core.oidc.endpoint.OidcParameterNames;
import org.springframework.security.oauth2.server.authorization.OAuth2TokenType;
import org.springframework.security.oauth2.server.authorization.token.JwtEncodingContext;
import org.springframework.security.oauth2.server.authorization.token.OAuth2TokenCustomizer;

import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@Configuration(proxyBeanMethods = false)
public class IdTokenCustomizerConfig {

    @Bean
    public OAuth2TokenCustomizer<JwtEncodingContext> tokenCustomizer() {
        return (context) -> {
            if (OidcParameterNames.ID_TOKEN.equals(context.getTokenType().getValue()) || OAuth2TokenType.ACCESS_TOKEN.equals(context.getTokenType())) {
                context.getClaims().claims(claims -> claims.putAll(buildCustomClaims(context)));
            }
        };
    }

    private Map<String, Object> buildCustomClaims(final JwtEncodingContext context) {
        final DefaultPrincipal principal = (DefaultPrincipal) context.getPrincipal().getPrincipal();

        final Set<String> roles = principal.getAuthorities()
                .stream().map(GrantedAuthority::getAuthority)
                .collect(Collectors.toSet());

        return Map.of("roles", roles, "user_id", principal.getId());
    }
}

File: src/main/kotlin/by/devpav/kotlin/oidcidp/config/IdTokenCustomizerConfig.kt
package by.devpav.kotlin.oidcidp.config

import by.devpav.kotlin.oidcidp.config.JwtClaims.USER_ROLES_CLAIM_NAME
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.oauth2.core.oidc.endpoint.OidcParameterNames
import org.springframework.security.oauth2.server.authorization.token.JwtEncodingContext
import org.springframework.security.oauth2.server.authorization.token.OAuth2TokenCustomizer
import java.util.stream.Collectors

@Configuration(proxyBeanMethods = false)
class IdTokenCustomizerConfig {

    @Bean
    fun tokenCustomizer(): OAuth2TokenCustomizer<JwtEncodingContext> {
        return OAuth2TokenCustomizer { context ->
            when {
                (OidcParameterNames.ID_TOKEN == context.tokenType.value) -> {
                    context.claims.claims { claims ->
                        claims[USER_ROLES_CLAIM_NAME] = context.getPrincipal<Authentication>().authorities
                            .stream()
                            .map { obj: GrantedAuthority -> obj.authority }
                            .collect(Collectors.toSet())
                    }
                }
            }
        }
    }

}
